### ISSUE TITLE
Over-zealous username validation in UI

### ISSUE THREAD

**Issue by mrichar1** [go-gitea/gitea#4150]
- Gitea version (or commit ref): 1.4.1
- Git version: 1.8.3.1
- Operating system: Redhat EL7.4
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## Description
We use kerberos via apache to authenticate hosts to gitea. Hosts have kerberos principals of the form `git/host.example.com@EXAMPLE.COM` which get translated to the username `git/host.example.com`

When we set `ENABLE_REVERSE_PROXY_AUTO_REGISTRATION` to `true` then these accounts are correctly created in gitea when the host runs git commands against our server. However since the account is 'new' it is not in the appropriate teams, so the first run fails, and we then have to set up team membership etc and re-run the processes.

To fix this I'm trying to pre-create these accounts using the UI - however when I try to create a new user with name `git/host.example.com` I receive the error `Username must be valid alphanumeric, dash(-_) or dot characters.`

Since auto-generated usernames with these characters seem to work fine, would it be possible to alter this validation to allow a broader range of usernames, filtering only those which would make gitea actually fail?

**Comment by lafriks** [comment]
`/` can not be allowed as that could break functionality and even rise security issues

**Comment by gszy** [comment]
What’s your opinion on allowing `@` in usernames?

**Comment by lafriks** [comment]
Also quite risky to break something. Would all git clients support such urls?

**Comment by gszy** [comment]
TL; DR: don’t know…

`git check-ref-format --branch 'a@a'` returns 0, but `man git-push` doesn’t explain what characters are allowed in `path/to/repo` (in `ssh://[user@]host.xz[:port]/path/to/repo.git/` example). Neither GitHub nor GitLab allow `@` in usernames or project names. 

`@` is commonly used to mention a user or to refer to specific commit. Unless username’s first character is `@` or it looks like `me@something­‑that­‑looks­‑like­‑commit­‑SHA`, that shouldn’t be a very big problem, though it would be if we could refer to branches, like `me@master` (could be solved with a config option).

**Comment by lafriks** [comment]
I do think that we should stick to current behaviour to not break things and keep compatibility with future changes (especially if we implement projects/groups under organizations).

**Comment by cwchristerw** [comment]
I need "-" to work in org name when transferring repo between orgs

**Comment by silverwind** [comment]
I think we should align our validation for org and repo name to a common denominator among forges like GitHub/GitLab. GitHub for example allow `-` as repo name, but not as a org name.

The most accurate description of what is allowed in repos is [this](https://stackoverflow.com/a/59082561/808699):

> it looks like github allows [A-Za-z0-9_.-], and transforms all other characters to "-".

For orgs:

>  The name may only contain alphanumeric characters or single hyphens, and cannot begin or end with a hyphen.

**Comment by silverwind** [comment]
As per https://github.com/go-gitea/gitea/issues/35408, org names, contrary to user names can at least end in a `-` and also `--`, so likely "cannot begin or end with a hyphen" does not apply to org names, only to user names.
